### DISCUSSION TITLE
Can we improve the parallation of volume.tier.move?

### DISCUSSION THREAD

**Discussion by NyaMisty** [seaweedfs/seaweedfs#5383] [Ideas]
Currently the `volume.tier.move` can only parallel on each hdd server. However I believe a parallel can be achieved on volume level. Can we add this feature? Thanks!

**Comment by NyaMisty** [comment]
Hi @chrislusf I've found a bit clue about this.

It seems there's a strange Lock causing the slow speed: https://github.com/seaweedfs/seaweedfs/blob/77fd62f6a59a3f9389707886a9a4b414d20099ec/weed/shell/command_volume_tier_move.go#L125

I believe that's because when moving an volume, we assume that two servers has same R/W speed, so there's no need for concurrency.
However I've moving from ssd to hdd, and ssd is several times faster than hdd, I believe the concurrency is pretty needed.

Is thisi lock a **hard** limit or just a speed limit?

**Reply by chrislusf** [reply]
it is preventing multiple volumes moving out of the same volume server, to avoid overloading one volume server.

**Reply by NyaMisty** [reply]
yeah, do you mean it's not a hard limit if the source volume server is far better than destination server?

**Reply by NyaMisty** [reply]
and there's another problem in the volume.tier.move command:
https://github.com/seaweedfs/seaweedfs/blob/77fd62f6a59a3f9389707886a9a4b414d20099ec/weed/shell/command_volume_tier_move.go#L218

Here we rotate the volume server list in the parent function via `allLocations = rotateDataNodes(allLocations)`. Although the list is rotated, but there's only one channel sender. If I understand correctly, each queue channel has the some buffer so that this loop won't blocked too much. However it's common that different volume server cannot achieve same write speed, and these unbalance write speed will directly block this loop, drastically decrease the move speed.

Here's a example fix of above problems:
https://github.com/seaweedfs/seaweedfs/commit/93040b8240259833c1b53e7a0b34906f0a9474aa

It addressed theses issues:
1. Add an option to remove source volume server lock
2. Instead of directly sending task to channel, we test if the target queue is free, and only send task to free queues.

**Reply by NyaMisty** [reply]
Hi @chrislusf  Any idea / suggestion on this solution?

**Reply by chrislusf** [reply]
currently the limit of concurrent transfer per volume server is just one.
We can add a configurable limit later.
